lib/public/Util.php

```diff
@@ -173,10 +176,74 @@ public static function addStyle($application, $file = null) {
 	 * add a javascript file
 	 * @param string $application
 	 * @param string $file
+	 * @param string $afterAppId
 	 * @since 4.0.0
 	 */
-	public static function addScript($application, $file = null) {
-		\OC_Util::addScript($application, $file);
+	public static function addScript($application, $file = null, $afterAppId = null) {
+		if (!empty($application)) {
+			$path = "$application/js/$file";
+		} else {
+			$path = "js/$file";
+		}
+
+		// Inject js translations if we load a script for
+		// a specific app that is not core, as those js files
+		// need separate handling
+		if ($application !== 'core' && $file !== null) {
+			self::addTranslations($application);
+		}
+
+		// init app array if it doesn't exists
+		if (!array_key_exists($application, self::$scripts)) {
+			self::$scripts[$application] = ['first' => [], 'last' => []];
+		}
+
+		/**
+		 * manage priorities if defined
+		 * we store the data like this, then flatten everything
+		 * [
+		 * 'core' => [
+		 * 'first' => [
+		 * '/core/js/main.js',
+		 * ],
+		 * 'last' => [
+		 * '/apps/viewer/js/viewer-main.js',
+		 * ]
+		 * ],
+		 * 'viewer' => [
+		 * 'first' => [
+		 * '/apps/viewer/js/viewer-public.js',
+		 * ],
+		 * 'last' => [
+		 * '/apps/files_pdfviewer/js/files_pdfviewer-main.js',
+		 * ]
+		 * ]
+		 * ]
+		 */
+		if (!empty($afterAppId)) {
+			// init afterAppId app array if it doesn't exists
+			if (!array_key_exists($afterAppId, self::$scripts)) {
+				self::$scripts[$afterAppId] = ['first' => [], 'last' => []];
+			}
+			self::$scripts[$afterAppId]['last'][] = $path;
+		} else {
+			self::$scripts[$application]['first'][] = $path;
+		}
+	}
```

Comment:
Move the `self::$scripts[$application]` creation inside the if to match the `self::$scripts[$afterAppId]` creation flow?

```suggestion
		if (!empty($afterAppId)) {
			// init afterAppId app array if it doesn't exists
			if (!array_key_exists($afterAppId, self::$scripts)) {
				self::$scripts[$afterAppId] = ['first' => [], 'last' => []];
			}
			self::$scripts[$afterAppId]['last'][] = $path;
		} else {
			if (!array_key_exists($application, self::$scripts)) {
				self::$scripts[$application] = ['first' => [], 'last' => []];
			}
			self::$scripts[$application]['first'][] = $path;
		}
	}


		
```